pkg/sql/lease.go

```diff
@@ -1149,7 +1148,30 @@ func (m *LeaseManager) Release(lease *LeaseState) error {
 	// could be bad if a lot of tables keep being created. I looked into cleaning
 	// up a bit, but it seems tricky to do with the current locking which is split
 	// between LeaseManager and tableState.
-	return t.release(lease, m.LeaseStore)
+	return t.release(lease, m)
+}
+
+func (m *LeaseManager) isDrainingLeases() bool {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	return m.draining
+}
+
+// SetDraining (when called with 'true') removes all inactive leases. Any leases
+// that are active will be removed once the lease's reference count drops to 0.
+func (m *LeaseManager) SetDraining(drain bool) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.draining = drain
+	if !drain {
+		return
+	}
+	for _, t := range m.tables {
+		t.mu.Lock()
+		toRelease := append([]*LeaseState(nil), t.active.data...)
+		t.releaseLeasesIfNotActive(toRelease, m.LeaseStore)
+		t.mu.Unlock()
+	}
```

Comment:
Yes, the locking order is correct according to the comment on `LeaseManager` and it protects both `draining` and `m.tables`.